jobs/workers.go

```diff
@@ -53,171 +35,39 @@ func (srv *JobServer) InitWorkers() error {
-		workers.ElasticsearchIndexing = elasticsearchIndexerInterface.MakeWorker()
-	}
-
-	if elasticsearchAggregatorInterface := srv.ElasticsearchAggregator; elasticsearchAggregatorInterface != nil {
-		workers.ElasticsearchAggregation = elasticsearchAggregatorInterface.MakeWorker()
-	}
-
-	if ldapSyncInterface := srv.LdapSync; ldapSyncInterface != nil {
-		workers.LdapSync = ldapSyncInterface.MakeWorker()
-	}
-
-	if migrationsInterface := srv.Migrations; migrationsInterface != nil {
-		workers.Migrations = migrationsInterface.MakeWorker()
-	}
-
-	if pluginsInterface := srv.Plugins; pluginsInterface != nil {
-		workers.Plugins = pluginsInterface.MakeWorker()
-	}
-
-	if bleveIndexerInterface := srv.BleveIndexer; bleveIndexerInterface != nil {
-		workers.BleveIndexing = bleveIndexerInterface.MakeWorker()
-	}
-
-	if expiryNotifyInterface := srv.ExpiryNotify; expiryNotifyInterface != nil {
-		workers.ExpiryNotify = expiryNotifyInterface.MakeWorker()
-	}
-
-	if activeUsersInterface := srv.ActiveUsers; activeUsersInterface != nil {
-		workers.ActiveUsers = activeUsersInterface.MakeWorker()
-	}
-
-	if productNoticesInterface := srv.ProductNotices; productNoticesInterface != nil {
-		workers.ProductNotices = productNoticesInterface.MakeWorker()
-	}
-
-	if importProcessInterface := srv.ImportProcess; importProcessInterface != nil {
-		workers.ImportProcess = importProcessInterface.MakeWorker()
-	}
-
-	if importDeleteInterface := srv.ImportDelete; importDeleteInterface != nil {
-		workers.ImportDelete = importDeleteInterface.MakeWorker()
-	}
-
-	if exportProcessInterface := srv.ExportProcess; exportProcessInterface != nil {
-		workers.ExportProcess = exportProcessInterface.MakeWorker()
-	}
-
-	if exportDeleteInterface := srv.ExportDelete; exportDeleteInterface != nil {
-		workers.ExportDelete = exportDeleteInterface.MakeWorker()
-	}
-
-	if cloudInterface := srv.Cloud; cloudInterface != nil {
-		workers.Cloud = cloudInterface.MakeWorker()
-	}
+	workers.Watcher = srv.MakeWatcher(workers, DefaultWatcherPollingInterval)
 
-	if resendInvitationEmailInterface := srv.ResendInvitationEmails; resendInvitationEmailInterface != nil {
-		workers.ResendInvitationEmail = resendInvitationEmailInterface.MakeWorker()
-	}
+	srv.workers = workers
 
-	if extractContentInterface := srv.ExtractContent; extractContentInterface != nil {
-		workers.ExtractContent = extractContentInterface.MakeWorker()
-	}
+	return nil
+}
 
-	if fixCRTChannelUnreads := srv.FixCRTChannelUnreads; fixCRTChannelUnreads != nil {
-		workers.FixCRTChannelUnreads = fixCRTChannelUnreads.MakeWorker()
+func NewWorkers(configService configservice.ConfigService) *Workers {
+	return &Workers{
+		ConfigService: configService,
+		workers:       make(map[string]model.Worker),
 	}
+}
 
-	srv.workers = workers
+func (workers *Workers) AddWorker(name string, worker model.Worker) {
+	workers.workers[name] = worker
+}
 
-	return nil
+func (workers *Workers) Get(name string) model.Worker {
+	return workers.workers[name]
```

Comment:
Can this get called concurrently with `AddWorker()`? Do we need to guard it somehow?

Reply:
Actually, this shouldn't happen concurrently, the AddWorker calls is done at initialization time, but adding a lock here is possible if we feel like is needed (At least with the current implementation I don't see the need of that, but we can end adding something in the future that interferes with that).